components/cam/src/physics/cam/micro_p3.F90

```diff
@@ -1260,70 +1262,81 @@ SUBROUTINE p3_main(qc,nc,qr,nr,th_old,th,qv_old,qv,dt,qitot,qirim,nitot,birim,ss
           ! allow ice nucleation if < -15 C and > 5% ice supersaturation
           ! use CELL-AVERAGE values, freezing of vapor
 
-          if (t(i,k).lt.icenuct .and. supi(i,k).ge.0.05) then
+         if(.not. log_predictNc) then 
+            if ( t(i,k).lt.icenuct .and. supi(i,k).ge.0.05) then
 
-             !        dum = exp(-0.639+0.1296*100.*supi(i,k))*1000.*inv_rho(i,k)  !Meyers et al. (1992)
-             dum = 0.005*exp(0.304*(zerodegc-t(i,k)))*1000.*inv_rho(i,k)   !Cooper (1986)
-             dum = min(dum,100.e3*inv_rho(i,k))
-             N_nuc = max(0.,(dum-nitot(i,k))*odt)
+               !        dum = exp(-0.639+0.1296*100.*supi(i,k))*1000.*inv_rho(i,k)  !Meyers et al. (1992)
+               dum = 0.005*exp(0.304*(zerodegc-t(i,k)))*1000.*inv_rho(i,k)   !Cooper (1986)
+               dum = min(dum,100.e3*inv_rho(i,k))
+               N_nuc = max(0.,(dum-nitot(i,k))*odt)
 
-             if (N_nuc.ge.1.e-20) then
-                Q_nuc = max(0.,(dum-nitot(i,k))*mi0*odt)
-                qinuc = Q_nuc
-                ninuc = N_nuc
-             endif
+               if (N_nuc.ge.1.e-20) then
+                  Q_nuc = max(0.,(dum-nitot(i,k))*mi0*odt)
+                  qinuc = Q_nuc
+                  ninuc = N_nuc
+               endif
+
+            endif
+         endif 
 
-          endif
 
+         if(log_predictNc) then 
+         ! Ice nucleation predicted by aerosol scheme 
+            if  ( t(i,k) .lt. 258.15 .and. supi(i,k).ge.0.05) then
+               ninuc = max(0., (naai(i,k) - nitot(i,k))*odt)
+               qinuc = ninuc * mi0
```

Comment:
In micro_p3_utils, it says mi0=4.*piov3*900.*1.e-18 where piov3 is pi/3 => qinuc=number of new drops/kg * mass of each new drop assuming a radius of (900e-18)^(1/3)=9.65e-6... I'm not sure what the units are, but this seems relatively big. presumably, some of that 900e-18 is the conversion from drop volume to mass/volume of air and from volume of air to kg of air... but I just wanted to check that we're sure this mi0 term is correct...

Reply:
The expression is consistent with Appendix C section a of Morrison and Milbrandt (2015). The factor of 900 in the expression for mi0 is actually the density of ice 900 kgm-3 (note that Morrison and Milbrandt use 917kgm-3). The radius of the droplet is (1e-18)^-3  m = 1e-6 m or 1 micron.

Reply:
The expression is consistent with Appendix C section a of Morrison and Milbrandt (2015). The factor of 900 in the expression for mi0 is actually the density of ice 900 kgm-3 (note that Morrison and Milbrandt use 917kgm-3). The radius of the droplet is (1e-18)^-3 m = 1e-6 m or 1 micron.

Reply:
Awesome, thanks for checking!